Make rdf and oai usable on ingest, add service verification script

terrywbrady opened the pull request:
### Associated Issues
- https://github.com/DSpace-Labs/DSpace-Docker-Images/issues/100
- https://github.com/DSpace-Labs/DSpace-Docker-Images/issues/102
- https://github.com/DSpace-Labs/DSpace-Docker-Images/issues/101

## Multi-app Dockerfile with 

This uses prod dockerfile.  This uses the new docker-compose files that are set to work with one webapp.

_Note that BASEROOT is needed to override the assumed /spring-rest path._
```
$ BASEROOT=http://localhost:8080 docker-compose -p d7 -f docker-compose.yml -f d7.override.yml -f sword.override.yml -f oai.override.yml -f rdf.override.yml down                      R
```

### All Services work

```
$ BASEROOT=http://localhost:8080 ./testComposeConfigs.sh
  ---> Tomcat http://localhost:8080
  ---> Spring-Rest http://localhost:8080
  ---> Angular http://localhost:3000
  ---> Legacy-Rest http://localhost:8080/rest
  ---> OAI http://localhost:8080/oai/request?verb=Identify
  ---> RDF http://localhost:8080/rdf/handle/123456789/1/rdf?text=true
  ---> Triplestore http://localhost:3030
  ---> SOLR-External http://localhost:8983/solr
  ---> Sword v1 http://localhost:8080/sword/deposit/123456789/3
  ---> Sword v2 http://localhost:8080/swordv2/collection/123456789/3

```

## Onewebapp Dockerfile with /spring-rest and /rest

```
DOCKER_OWNER=terrywbrady docker-compose -p d7 -f docker-compose.yml -f d7.override.yml -f sword.override.yml -f oai.override.yml -f rdf.override.yml up -d      
```

This uses Dockerfile.jdk8-test from https://github.com/tdonohue/DSpace/pull/9

### RDF and Sword v1 do not work

```
$ ./testComposeConfigs.sh
  ---> Tomcat http://localhost:8080
  ---> Spring-Rest http://localhost:8080/spring-rest
  ---> Angular http://localhost:3000
  ---> Legacy-Rest http://localhost:8080/rest
  ---> OAI http://localhost:8080/spring-rest/oai/request?verb=Identify
  ---> RDF http://localhost:8080/spring-rest/rdf/handle/123456789/1/rdf?text=true
 *** ERROR RDF
HTTP/1.1 500
  ---> Triplestore http://localhost:3030
  ---> SOLR-External http://localhost:8983/solr
  ---> Sword v1 http://localhost:8080/spring-rest/sword/deposit/123456789/3
 *** ERROR Sword v1
HTTP/1.1 400
  ---> Sword v2 http://localhost:8080/spring-rest/swordv2/collection/123456789/3

```

## Onewebapp Dockerfile with /ROOT

DOCKER_OWNER=terrywbrady DSPACE_VER=one-webapp docker-compose -p d7 -f docker-compose.yml -f d7.override.yml -f sword.override.yml -f oai.override.yml -f rdf.override.yml up -d      

This uses Dockerfile.onewebapp.jdk8 from https://github.com/tdonohue/DSpace/pull/9

### RDF and Legacy REST Fail

_Legacy REST failure is expected_

```
$ BASEROOT=http://localhost:8080 ./testComposeConfigs.sh
  ---> Tomcat http://localhost:8080
  ---> Spring-Rest http://localhost:8080
  ---> Angular http://localhost:3000
  ---> Legacy-Rest http://localhost:8080/rest
 *** ERROR Legacy-Rest
HTTP/1.1 404
  ---> OAI http://localhost:8080/oai/request?verb=Identify
  ---> RDF http://localhost:8080/rdf/handle/123456789/1/rdf?text=true
 *** ERROR RDF
HTTP/1.1 500
  ---> Triplestore http://localhost:3030
  ---> SOLR-External http://localhost:8983/solr
  ---> Sword v1 http://localhost:8080/sword/deposit/123456789/3
  ---> Sword v2 http://localhost:8080/swordv2/collection/123456789/3

```

tdonohue commented on docker-compose-files/dspace-compose/sword.override.yml:
```diff
@@ -4,12 +4,11 @@ version: "3.7"
-      - swordv2__D__server__P__servicedocument__P__url=http://localhost:8080/swordv2/servicedocument
+      - swordv2__D__server__P__url=http://localhost:8080/spring-rest/swordv2
+      - swordv2__D__server__P__collection__P__url=http://localhost:8080/spring-rest/swordv2/collection
+      - swordv2__D__server__P__servicedocument__P__url=http://localhost:8080/swordv2/spring-rest/servicedocument
```
@terrywbrady : Just a sidenote, if these three URL configurations are necessary for SWORDv2, then I suspect they **also** will be necessary for SWORDv1.  Both SWORD webapps do their own URL validation, and they actively throw 404 errors if a request doesn't match the expected URL *exactly*.  This is something I might be able to cleanup (with more sane default URLs for SWORDv1/v2) in the follow-up PR to https://github.com/DSpace/DSpace/pull/2265, as that follow-up PR will better standardize URLs to all be based off a (new) `dspace.server.url` path.  But, for now, we may need to configure the proper URLs for SWORDv1 and SWORDv2 to work as expected.

terrywbrady commented:
@tdonohue , I did some more testing and updated the description for this PR using an improved test driver script.  I have now modified the script to allow you to specify the basePath for the web services: localhost:8080/spring-rest or localhost:8080

Note that with production docker images (master), all services succeed.

With the one webapp PR + my changes to the Dockerfiles, I see the following
- RDF always fails
- Sword v1 fails when the baseUrl is at localhost:8080/spring-rest

Based on the differing results, I do not think I am setting variables incorrectly, but I would like to confirm this with you.

tdonohue commented:
@terrywbrady : Is there any way to get the error logs from both RDF and SWORDv1 when the failure occurs?  At a glance, I'm not seeing an obvious configuration problem, but if we could get more information on the error being thrown by RDF & SWORDv1, that might help figure it out.  I am surprised that SWORDv2 works but SWORDv1 doesn't (as they are quite similar in their setup / configuration), but again the error logs might make it obvious where the problem is.  

It does sound more likely that this may not be a configuration issue -- maybe you've hit upon a few bugs in the one webapp PR.  I will admit, I did *not* do extensive testing of RDF locally (I was more dependent on the ITs reporting issues).  I did test the SWORDv1 interface though and thought it was working properly -- though maybe a minor bug snuck in there at the last minute.

tdonohue commented on tools/testComposeConfigs.sh:
```diff
@@ -0,0 +1,45 @@
+# checkUrl "SOLR-Statistics" ${SOLR}/authority/select?q=*%3A* 200
+
+echo "  ---> Sword v1 ${BR}/sword/deposit/123456789/3"
+curl -i --data-binary "@sword/example.zip" -H "Content-Disposition:filename=sword/example.zip" -H "Content-Type:application/zip" -H "X-Packaging:http://purl.org/net/sword-types/METSDSpaceSIP" -u test@test.edu:admin ${BR}/sword/deposit/123456789/4 2>&1 | grep "^HTTP" > /tmp/out.txt
```
@terrywbrady : small note. It's likely just a typo, and not a bug.  The SWORDv1 deposit is being sent to "123456789/4" but the `echo` in the previous line claims it is going to "123456789/3".

In the `testSword.sh` script, I see it also submits to "123456789/4", so I'm assuming that's likely the correct path?

tdonohue replied:
Also, I'm sorta surprised this `curl` command (and the one in `testSword.sh` does *not* have a `-X POST`.  I'm assuming it must be defaulting to POST though, as I think these deposits only work via POST.  So, again, likely not a bug, but possibly a minor improvement.

terrywbrady commented:
@tdonohue , here is the RDF issue that I see.  Note: `dspace rdfizer -c -v` runs successfully.

```
2019-03-28 15:19:13,286 INFO  org.springframework.cache.ehcache.EhCacheManagerFactoryBean @ Shutting down EhCache CacheManager
2019-03-28 15:19:18,163 ERROR org.springframework.boot.web.support.ErrorPageFilter @ Forwarding to error page from request [/rdf/handle/123456789/1/rdf] due to exception [javax.servl
et.ServletException: Servlet execution threw an exception]
javax.servlet.ServletException: javax.servlet.ServletException: Servlet execution threw an exception
        at org.dspace.utils.servlet.DSpaceWebappServletFilter.doFilter(DSpaceWebappServletFilter.java:91) ~[dspace-services-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:208) ~[spring-security-web-4.1.4.RELEASE.jar:4.1.4.RELEASE]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177) ~[spring-security-web-4.1.4.RELEASE.jar:4.1.4.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:105) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:81) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:197) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:117) [spring-boot-1.4.4.RELEASE.jar:1.4.4.RELEASE]
        at org.springframework.boot.web.support.ErrorPageFilter.access$000(ErrorPageFilter.java:61) [spring-boot-1.4.4.RELEASE.jar:1.4.4.RELEASE]
        at org.springframework.boot.web.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:92) [spring-boot-1.4.4.RELEASE.jar:1.4.4.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) [spring-web-4.3.6.RELEASE.jar:4.3.6.RELEASE]
        at org.springframework.boot.web.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:110) [spring-boot-1.4.4.RELEASE.jar:1.4.4.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.38]
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [log4j-web-2.6.2.jar:2.6.2]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) [catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) [catalina.jar:8.5.38]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199) [catalina.jar:8.5.38]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [catalina.jar:8.5.38]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:493) [catalina.jar:8.5.38]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140) [catalina.jar:8.5.38]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81) [catalina.jar:8.5.38]
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:650) [catalina.jar:8.5.38]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87) [catalina.jar:8.5.38]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342) [catalina.jar:8.5.38]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:800) [tomcat-coyote.jar:8.5.38]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-coyote.jar:8.5.38]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806) [tomcat-coyote.jar:8.5.38]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498) [tomcat-coyote.jar:8.5.38]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-coyote.jar:8.5.38]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_181]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_181]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-util.jar:8.5.38]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
Caused by: javax.servlet.ServletException: Servlet execution threw an exception
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:238) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) ~[tomcat-websocket.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.dspace.app.rest.filter.DSpaceRequestContextFilter.doFilter(DSpaceRequestContextFilter.java:45) ~[classes/:7.0-SNAPSHOT]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.dspace.utils.servlet.DSpaceWebappServletFilter.doFilter(DSpaceWebappServletFilter.java:84) ~[dspace-services-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        ... 47 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.apache.jena.riot.web.HttpOp
        at org.apache.jena.web.DatasetGraphAccessorHTTP.doGet(DatasetGraphAccessorHTTP.java:134) ~[jena-arq-2.13.0.jar:2.13.0]
        at org.apache.jena.web.DatasetGraphAccessorHTTP.httpGet(DatasetGraphAccessorHTTP.java:128) ~[jena-arq-2.13.0.jar:2.13.0]
        at org.dspace.rdf.storage.RDFStorageImpl.load(RDFStorageImpl.java:62) ~[dspace-api-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        at org.dspace.rdf.RDFUtil.loadModel(RDFUtil.java:103) ~[dspace-api-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        at org.dspace.rdf.providing.DataProviderServlet.serveNamedGraph(DataProviderServlet.java:137) ~[dspace-rdf-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        at org.dspace.rdf.providing.DataProviderServlet.processRequest(DataProviderServlet.java:129) ~[dspace-rdf-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        at org.dspace.rdf.providing.DataProviderServlet.doGet(DataProviderServlet.java:221) ~[dspace-rdf-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        at javax.servlet.http.HttpServlet.doHead(HttpServlet.java:245) ~[servlet-api.jar:?]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[servlet-api.jar:?]
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:742) ~[servlet-api.jar:?]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52) ~[tomcat-websocket.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.dspace.app.rest.filter.DSpaceRequestContextFilter.doFilter(DSpaceRequestContextFilter.java:45) ~[classes/:7.0-SNAPSHOT]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[catalina.jar:8.5.38]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[catalina.jar:8.5.38]
        at org.dspace.utils.servlet.DSpaceWebappServletFilter.doFilter(DSpaceWebappServletFilter.java:84) ~[dspace-services-7.0-SNAPSHOT.jar:7.0-SNAPSHOT]
        ... 47 more

```

terrywbrady commented:
@tdonohue , now sword v1 is working for me when I deploy to ROOT.  When it is deployed to spring-rest, I get the following error.

```
$ curl -i --data-binary "@sword/example.zip" -H "Content-Disposition:filename=sword/example.zip" -H "Content-Type:application/zip" -H "X-Packaging:http://purl.org/net/sword-types/METSDSpaceSIP" -u test@test.edu:admin -X POST http://localhost:8080/spring-rest/sword/deposit/123456789/4 2>&1
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 34160    0   383  100 33777   3481   299k --:--:-- --:--:-- --:--:--  303kHTTP/1.1 100

HTTP/1.1 400
Content-Type: application/atom+xml;charset=UTF-8
Transfer-Encoding: chunked
Date: Thu, 28 Mar 2019 15:37:43 GMT
Connection: close

<sword:error xmlns:sword="http://purl.org/net/sword/" xmlns:atom="http://www.w3.org/2005/Atom" href="http://www.dspace.org/ns/sword/1.3.1/errors/BadUrl"><atom:summary>The deposit URL does not resolve to a valid deposit target</atom:summary><atom:title>ERROR</atom:title><atom:updated>2019-03-28T15:37:43.293Z</atom:updated><sword:userAgent>curl/7.58.0</sword:userAgent></sword:error>
```

tdonohue commented:
@terrywbrady : Thanks for those error details! That is extremely helpful.

* The RDF error sounds like a dependency problem. I'll look into why that class isn't coming across properly...it's possible I made a dependency mistake in one of my POM changes.  So, this definitely sounds like a dependency bug in my "one webapp" PR.
* The SWORD error is the SWORD URL validation code causing havoc -- the error displayed is saying the URL you posted to isn't matching what SWORD expects (it's very strict, expecting an exact string match). I *know* I will have to refactor this code in the follow-up PR (when I introduce the `dspace.server.url` field).  So, this seems like something we could consider logging as a "known bug" and assigning me to fix it in the follow-up PR. It's good to hear though that the functionality works fine when deployed as ROOT. I can dig further into what might be going on when deployed on a subpath.

terrywbrady commented:
I tested these docker compose changes with DSpace 5, 6, and 7.  The changes seem to work well.  Merging...
